packages/sit-onyx/src/components/OnyxDataGrid/features/index.ts

```diff
@@ -0,0 +1,178 @@
+};
+
+/**
+ * Helper function that infers the generics of the TableFeature type.
+ * @example
+ * ```ts
+ *
+ * const MY_FEATURE = Symbol("TABLE_HEADER_BUTTON");
+ * export const useTableHeaderButton = createFeature(<TEntry extends DataGridEntry>() => {
+ *   return {
+ *     name: MY_FEATURE,
+ *     header: {
+ *       actions: (column) => [
+ *         {
+ *           iconComponent: h('button', { onClick: (column) => console.log(`Clicked on ${column}`) },),
+ *         },
+ *       ],
+ *     },
+ *   };
+ * });
+ * ```
+ */
+export const createFeature = <TFeatureName extends symbol, TArgs extends unknown[]>(
+  featureDefinition: <TEntry extends DataGridEntry>(
+    ...args: TArgs
+  ) => TableFeature<TEntry, TFeatureName>,
+) => featureDefinition;
+
+type ExtractTEntry<T> = T extends TableFeature<infer I, symbol>[] ? I : never;
+
+/**
+ * Uses the defined table features to provide factory functions.
+ * These factories are to be used to map data and configuration to `OnyxDataGridRenderer` properties.
+ * The properties are then used to render the data grid.
+ *
+ * Make use of the `watchSources` to trigger re-rendering when state changes occur.
+ * @example
+ * ```vue
+ * <script setup lang="ts">
+ * // ...
+ * // imports, props, emits, etc.
+ * const withHeaderButton = useTableHeaderButton<TEntry>();
+ *
+ * const { watchSources, createRendererRows, createRendererColumns } = useTableFeatures([withSorting]);
+ *
+ * const renderCols: Ref<DataGridRendererColumn<TEntry, object>[]> = ref([]);
+ * const renderRows: Ref<DataGridRendererRow<TEntry, DataGridMetadata>[]> = ref([]);
+ *
+ * const { columns, data } = toRefs(props);
+ *
+ * watch(
+ *   [columns, data, ...watchSources],
+ *   ([newColumns, newData]) => {
+ *     renderCols.value = createRendererColumns(newColumns);
+ *     renderRows.value = createRendererRows(newData, newColumns);
+ *   },
+ *   { immediate: true },
+ * );
+ * </script>
+ * <template>
+ *   <OnyxDataGridRenderer :columns="renderCols" :rows="renderRows" />
+ * </template>
+ * ```
+ */
+export const useTableFeatures = <
+  T extends TableFeature<TEntry, symbol>[] | [],
+  TEntry extends DataGridEntry = ExtractTEntry<T>,
+>(
+  features: T,
+) => {
+  const createRendererColumns = (
+    columns: (keyof TEntry)[],
+  ): DataGridRendererColumn<TEntry, object>[] => {
+    const headerFeatures = features.map((feature) => feature.header).filter((header) => !!header);
+    const headerActions = headerFeatures
+      .map((feature) => feature.actions)
+      .filter((actions) => !!actions);
+
+    return columns.map((column) => {
+      const iconActions = headerActions
+        .flatMap((actionFactory) => actionFactory(column))
+        .map(({ iconComponent }) => iconComponent);
+
+      return {
+        key: column,
+        component: () =>
+          h("div", { class: getBemClass("header-cell") }, [
+            String(column),
+            ...iconActions.map((a) => h(a)),
+          ]),
+        props: {},
+      };
```

Comment:
Not sure if I like this inline component  + CSS in JS...what do you think about creating a .vue component for this?